File: GDTMobApp/Release/GDTAds/GDTNativeExpressAd/NativeExpressAdViewController.m
//
//  NativeExpressAdViewController.m
//  GDTMobApp
//
//  Created by michaelxing on 2017/4/17.
//  Copyright © 2017年 Tencent. All rights reserved.
//

#import "NativeExpressAdViewController.h"
#import "GDTNativeExpressAd.h"
#import "GDTNativeExpressAdView.h"
#import "GDTAppDelegate.h"
#import "NativeExpressAdConfigView.h"
#import "DemoUtil.h"

@interface NativeExpressAdViewController ()<GDTNativeExpressAdDelegete,UITableViewDelegate,UITableViewDataSource>

@property (nonatomic, strong) NSMutableArray *expressAdViews;

@property (nonatomic, strong) GDTNativeExpressAd *nativeExpressAd;

@property (weak, nonatomic) IBOutlet UITextField *placementIdTextField;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property (assign, nonatomic)  float widthSliderValue;
@property (assign, nonatomic)  float heightSliderValue;
@property (assign, nonatomic)  float adCountSliderValue;

//切换广告样式

@property (nonatomic, strong) UIButton *changAdvStyleButton;

@property (nonatomic, copy) NSString *token;
@property (nonatomic, assign) BOOL useToken;

@end

@implementation NativeExpressAdViewController

static NSString *ABOVEPH_BLOWTEXT_STR = @"2011313603627732";
static NSString *ABOVETEXT_BLOW_PH_STR = @"5018760231881135";
static NSString *LEFTPH_RIHGTTEXT_STR = @"1001317613724659";
static NSString *LEFTTEXT_RIGHTPH_STR = @"3061112693227741";
static NSString *TWOPH_AND_TEXT_STR = @"6001113663823874";
static NSString *ABOVEPH_BLOWTEXT_SURFACE_STR = @"6021015693625724";
static NSString *ABOVETEXT_SURFACE_BLOWPH_STR = @"2021617633021755";
static NSString *TEXTSURFACE_ONEPHOTO_STR = @"7011916673826880";
static NSString *HORIZONTAL_IMAGE_STR = @"8031910693627892";
static NSString *VERTICAL_IMAGE_STR = @"6021618693720821";
static NSString *THREE_SMALLPH_STR = @"1098462271285109";
static NSString *Mediator_STR = @"100014";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.automaticallyAdjustsScrollViewInsets = NO;

    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"nativeexpresscell"];
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"splitnativeexpresscell"];
    
    [self initVideoConfig];
    [self refreshButton:nil];
}

- (IBAction)selectADVStyle:(id)sender {
    UIAlertController *advStyleAlertController = [UIAlertController alertControllerWithTitle:@"请选择需要的广告样式" message:nil preferredStyle:[[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad ? UIAlertControllerStyleAlert : UIAlertControllerStyleActionSheet];
    NSArray *advTypeTextArray = [self getAdvTypeTextArray];
    
    for (NSInteger i = 0; i < advTypeTextArray.count; i++) {
        UIAlertAction *advTypeAction = [UIAlertAction actionWithTitle:advTypeTextArray[i][0]
                                                                style:UIAlertActionStyleDefault
                                                              handler:^(UIAlertAction * _Nonnull action) {
            self.placementIdTextField.placeholder = advTypeTextArray[i][1];
            [self refreshViewWithNewPosID];
        }];
        [advStyleAlertController addAction:advTypeAction];
    }
    [advStyleAlertController addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:nil]];
    if (advStyleAlertController.popoverPresentationController) {
        [advStyleAlertController.popoverPresentationController setPermittedArrowDirections:0];//去掉arrow箭头
        advStyleAlertController.popoverPresentationController.sourceView=self.view;
        advStyleAlertController.popoverPresentationController.sourceRect=CGRectMake(0, self.view.bounds.size.height, self.view.bounds.size.width, self.view.bounds.size.height);
    }
    [self presentViewController:advStyleAlertController
                       animated:YES
                     completion:nil];
}

- (NSArray *)getAdvTypeTextArray
{
    return @[ @[@"上图下文(图片尺寸1280×720)", ABOVEPH_BLOWTEXT_STR],
              @[@"上文下图(图片尺寸1280×720)", ABOVETEXT_BLOW_PH_STR],
              @[@"左图右文(图片尺寸1200×800)", LEFTPH_RIHGTTEXT_STR],
              @[@"左文右图(图片尺寸1200×800)", LEFTTEXT_RIGHTPH_STR],
              @[@"双图双文-图文(大图尺寸1280×720)", TWOPH_AND_TEXT_STR],
              @[@"文字浮层(上图下文1280×720)", ABOVEPH_BLOWTEXT_SURFACE_STR],
              @[@"文字浮层(上文下图1280×720)", ABOVETEXT_SURFACE_BLOWPH_STR],
              @[@"文字浮层(单图1280×720)", TEXTSURFACE_ONEPHOTO_STR],
              @[@"横版纯图片(图片尺寸1280×720)", HORIZONTAL_IMAGE_STR],
              @[@"竖版纯图片(图片尺寸800×1200)", VERTICAL_IMAGE_STR],
              @[@"三小图双文(图片尺寸228×150)", THREE_SMALLPH_STR],
             @[@"流量分配",Mediator_STR]
            ];
}

- (void)refreshViewWithNewPosID {
    [self.expressAdViews removeAllObjects];
    [self.tableView reloadData];
    
    NSString *placementId = self.placementIdTextField.text.length > 0? self.placementIdTextField.text: self.placementIdTextField.placeholder;
    if (self.useToken) {
        self.nativeExpressAd = [[GDTNativeExpressAd alloc] initWithPlacementId:placementId token:self.token
                                                                        adSize:CGSizeMake(self.widthSliderValue, self.heightSliderValue)];
    } else {
        self.nativeExpressAd = [[GDTNativeExpressAd alloc] initWithPlacementId:placementId
                                                                        adSize:CGSizeMake(self.widthSliderValue, self.heightSliderValue)];
    }
    self.nativeExpressAd.delegate = self;
    [self.nativeExpressAd loadAd:(NSInteger)self.adCountSliderValue];
}


- (IBAction)refreshButton:(id)sender {
    [self.expressAdViews removeAllObjects];
    [self.tableView reloadData];
    
    
    NSString *placementId = self.placementIdTextField.text.length > 0? self.placementIdTextField.text: self.placementIdTextField.placeholder;
    if (self.useToken) {
        self.nativeExpressAd = [[GDTNativeExpressAd alloc] initWithPlacementId:placementId token:self.token
                                                                        adSize:CGSizeMake(self.widthSliderValue, self.heightSliderValue)];
    } else {
        self.nativeExpressAd = [[GDTNativeExpressAd alloc] initWithPlacementId:placementId
                                                                        adSize:CGSizeMake(self.widthSliderValue, self.heightSliderValue)];
    }
    self.nativeExpressAd.delegate = self;
    [self.nativeExpressAd loadAd:(NSInteger)self.adCountSliderValue];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [super touchesBegan:touches withEvent:event];
    [self.view endEditing:YES];
}

- (void)initVideoConfig
{
    self.widthSliderValue = [UIScreen mainScreen].bounds.size.width;
    self.heightSliderValue = 50;
    self.adCountSliderValue = 3;
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"更多视频配置" style:UIBarButtonItemStylePlain target:self action:@selector(jumpToAnotherView)];
}

- (void)jumpToAnotherView{
    [self.navigationItem.rightBarButtonItem setEnabled:NO];
    NativeExpressAdConfigView *theNativeExpressAdConfigView = [[NativeExpressAdConfigView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height)];
    __weak typeof(self) _weakSelf = self;
    theNativeExpressAdConfigView.placementId = self.placementIdTextField.text.length > 0? self.placementIdTextField.text: self.placementIdTextField.placeholder;;
    theNativeExpressAdConfigView.widthSlider.value = self.widthSliderValue;
    theNativeExpressAdConfigView.heightSlider.value = self.heightSliderValue;
    theNativeExpressAdConfigView.adCountSlider.value = self.adCountSliderValue;
    theNativeExpressAdConfigView.useTokenSwitch.on = self.useToken;
    theNativeExpressAdConfigView.tokenLabel.text = self.token;
    theNativeExpressAdConfigView.callBackBlock = ^(float widthSliderValue,float heightSliderValue,float adCountSliderValue,BOOL navigationRightButtonIsenabled){
        [self.navigationItem.rightBarButtonItem setEnabled:navigationRightButtonIsenabled];
        _weakSelf.widthSliderValue = widthSliderValue;
        _weakSelf.heightSliderValue = heightSliderValue;
        _weakSelf.adCountSliderValue = adCountSliderValue;
    };
    theNativeExpressAdConfigView.tokenBlock = ^(BOOL useToken, NSString * _Nonnull token) {
        _weakSelf.useToken = useToken;
        _weakSelf.token = token;
    };
    [theNativeExpressAdConfigView showInView:self.view];
}

/**
 * 上报给优量汇服务端在开发者客户端竞价中优量汇的竞价结果，以便于优量汇服务端调整策略提供给开发者更合理的报价
 *
 * 优量汇竞价失败调用 biddingLoss，并填入优量汇竞败原因（必填）、竞胜ADN ID（选填）、竞胜ADN报价（选填）
 * 优量汇竞价胜出调用 biddingWin，并填入开发者期望扣费价格（单位分）和最高竞败出价（单位分）
 * 请开发者如实上报相关参数，以保证优量汇服务端能根据相关参数调整策略，使开发者收益最大化
*/

- (void)reportBiddingResult {
    if (self.useToken) {
        // 针对本次曝光的媒体期望扣费，常用扣费逻辑包括一价扣费与二价扣费
        // 当采用一价扣费时，胜者出价即为本次扣费价格；当采用二价扣费时，第二名出价为本次扣费价格；
        // 自己根据实际情况设置
        [[self.expressAdViews firstObject] setBidECPM:100];
    }
    else {
        NSInteger fakeWinPrice = 200;
        NSInteger fakeHighestPrice = 100;
        GDTAdBiddingLossReason fakeLossReason = GDTAdBiddingLossReasonLowPrice;
        NSString *fakeAdnId = @"WinAdnId";
        NSNumber *flag = [[NSUserDefaults standardUserDefaults] objectForKey:@"debug_setting_bidding_report"];
        NSInteger reportFlag = flag ? [flag integerValue] : 1;
        if (reportFlag == 1) {
            for (GDTNativeExpressAdView *iter in self.expressAdViews) {
                [iter sendWinNotificationWithInfo:@{GDT_M_W_E_COST_PRICE: @(fakeWinPrice), GDT_M_W_H_LOSS_PRICE: @(fakeHighestPrice)}];
            }
            
        } else if (reportFlag == 2) {
            for (GDTNativeExpressAdView *iter in self.expressAdViews) {
                [iter sendLossNotificationWithInfo:@{GDT_M_L_WIN_PRICE: @(fakeWinPrice), GDT_M_L_LOSS_REASON:@(fakeLossReason), GDT_M_ADNID: fakeAdnId}];
            }
        }
    }
}

#pragma mark - GDTNativeExpressAdDelegete
/**
 * 拉取广告成功的回调
 */
- (void)nativeExpressAdSuccessToLoad:(GDTNativeExpressAd *)nativeExpressAd views:(NSArray<__kindof GDTNativeExpressAdView *> *)views
{
    NSLog(@"%s",__FUNCTION__);
    self.expressAdViews = [NSMutableArray arrayWithArray:views];
    if (self.expressAdViews.count) {
        [self.expressAdViews enumerateObjectsUsingBlock:^(GDTNativeExpressAdView *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            GDTNativeExpressAdView *expressView = (GDTNativeExpressAdView *)obj;
            expressView.controller = self;
            if ([expressView isAdValid]) {
                [expressView render];
            }
            NSLog(@"extraInfo: %@", obj.extraInfo);
            NSLog(@"eCPM:%ld eCPMLevel:%@", [expressView eCPM], [expressView eCPMLevel]);
        }];
    }
    [self.tableView reloadData];
    
    // 在 bidding 结束之后, 调用对应的竞胜/竞败接口
    [self reportBiddingResult];
}

/**
 * 拉取广告失败的回调
 */
- (void)nativeExpressAdRenderFail:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

/**
 * 拉取原生模板广告失败
 */
- (void)nativeExpressAdFailToLoad:(GDTNativeExpressAd *)nativeExpressAd error:(NSError *)error
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"Express Ad Load Fail : %@",error);
}

- (void)nativeExpressAdViewRenderSuccess:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
    [self.tableView reloadData];
}

/**
 * 模板渲染失败的回调
 */
- (void)nativeExpressAdViewRenderFail:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdViewClicked:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdViewClosed:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
    [self.expressAdViews removeObject:nativeExpressAdView];
    [self.tableView reloadData];
}

- (void)nativeExpressAdViewExposure:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdView:(GDTNativeExpressAdView *)nativeExpressAdView playerStatusChanged:(GDTMediaPlayerStatus)status {
    NSString *statusString = [DemoUtil videoPlayerStatusStringFromStatus:status];
    NSLog(@"%s-----status:%@",__FUNCTION__,statusString);
}

- (void)nativeExpressAdViewWillPresentScreen:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdViewDidPresentScreen:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdViewWillDismissScreen:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

- (void)nativeExpressAdViewDidDismissScreen:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"%s",__FUNCTION__);
}

/**
 * 详解:当点击应用下载或者广告调用系统程序打开时调用
 */
- (void)nativeExpressAdViewApplicationWillEnterBackground:(GDTNativeExpressAdView *)nativeExpressAdView
{
    NSLog(@"--------%s-------",__FUNCTION__);
}

/**
 *  投诉成功回调
 */
- (void)gdtAdComplainSuccess:(id)ad {
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告投诉成功");
}

#pragma mark - UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row % 2 == 0) {
        UIView *view = [self.expressAdViews objectAtIndex:indexPath.row / 2];
        return view.bounds.size.height;
    }
    else {
        return 44;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.expressAdViews.count * 2;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = nil;
    if (indexPath.row % 2 == 0) {
        cell = [self.tableView dequeueReusableCellWithIdentifier:@"nativeexpresscell" forIndexPath:indexPath];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        
        UIView *subView = (UIView *)[cell.contentView viewWithTag:1000];
        if ([subView superview]) {
            [subView removeFromSuperview];
        }
        GDTNativeExpressAdView *view = [self.expressAdViews objectAtIndex:indexPath.row / 2];
        view.tag = 1000;
        [cell.contentView addSubview:view];
        
        cell.accessibilityIdentifier = @"nativeTemp_even_ad";
    } else {
        cell = [self.tableView dequeueReusableCellWithIdentifier:@"splitnativeexpresscell" forIndexPath:indexPath];
        cell.backgroundColor = [UIColor grayColor];
        cell.accessibilityIdentifier = @"nativeTemp_odd_ad";
    }
    return cell;
}

@end

File: GDTMobSample-Swift/GDTAdTypes/NativeExpressAdViewController/NativeExpressAdViewController.swift
//
//  NativeExpressAdViewController.swift
//  GDTMobSample-Swift
//
//  Created by nimomeng on 2018/8/15.
//  Copyright © 2018 Tencent. All rights reserved.
//

import UIKit

class NativeExpressAdViewController: UIViewController,UITableViewDelegate,UITableViewDataSource,GDTNativeExpressAdDelegete {

    private var expressAdViews:Array<GDTNativeExpressAdView>!
    private var nativeExpressAd:GDTNativeExpressAd!
    
    @IBOutlet weak var placementIdTextField: UITextField!
    @IBOutlet weak var tableView: UITableView!

    // 初始值
    var widthSliderValue = UIScreen.main.bounds.size.width
    var heightSliderValue : CGFloat = 50
    var adCountSliderValue = 3
    let ABOVEPH_BLOWTEXT_STR = "5030722621265924"

    let ABOVETEXT_BLOW_PH_STR = "8010090333885456"

    let LEFTPH_RIHGTTEXT_STR = "1080793303881448"

    let LEFTTEXT_RIGHTPH_STR = "9050097313684512"

    let TWOPH_AND_TEXT_STR = "5070297373087567"

    let WIDTHPHOTO_STR = "5070791337820394"

    let HEIGHTPHOTO_STR = "6090492353182599"

    let THREE_SMALLPH_STR = "9050492343889611"

    let ABOVETEXT_SURFACE_BLOWPH_STR = "9010495393982624"

    let ABOVEPH_BLOWTEXT_SURFACE_STR = "6020493353488605"

    let TEXTSURFACE_ONEPHOTO_STR = "3030690323789618"

    let HEIGHTERPHOTO_STR = "6060290383380659"

    let ADVTYPE_COUNT = 12
    
   

    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.contentInsetAdjustmentBehavior = .never
        
      
        tableView.delegate = self
        tableView.dataSource = self
        
        tableView.register(object_getClass(UITableViewCell()), forCellReuseIdentifier: "nativeexpresscell")
        tableView.register(object_getClass(UITableViewCell()), forCellReuseIdentifier: "splitnativeexpresscell")
        configAlertController()
        
    }
    func configAlertController()->UIAlertController{
        let changePosIdController = UIAlertController(title: "请选择广告类型", message: nil, preferredStyle: UIAlertControllerStyle.actionSheet)
        
        let advTypeTextArray : [String] = ["上图下文(图片尺寸1280×720)",
                                             "上文下图(图片尺寸1280×720)",
                                             "左图右文(图片尺寸1200×800)",
                                             "左文右图(图片尺寸1200×800)",
                                             "双图双文(大图尺寸1280×720)",
                                             "纯图片(图片尺寸1280×720)",
                                             "纯图片(图片尺寸800×1200)",
                                             "三小图(图片尺寸228×150)",
                                             "文字浮层(上文下图1280×720)",
                                             "文字浮层(上图下文1280×720)",
                                             "文字浮层(单图1280×720)",
                                             "纯图片(图片尺寸1080×1920或800×1200)"]
            
        
        
        let placeholderArray : [String] = [ABOVEPH_BLOWTEXT_STR,
                                ABOVETEXT_BLOW_PH_STR,
                                           LEFTPH_RIHGTTEXT_STR,
                                           LEFTTEXT_RIGHTPH_STR,
                                           TWOPH_AND_TEXT_STR,
                                           WIDTHPHOTO_STR,
                                           HEIGHTPHOTO_STR,
                                           THREE_SMALLPH_STR,
                                           ABOVETEXT_SURFACE_BLOWPH_STR,
                                           ABOVEPH_BLOWTEXT_SURFACE_STR,
                                           TEXTSURFACE_ONEPHOTO_STR,
                                           HEIGHTERPHOTO_STR]
            for i in Range(0..<ADVTYPE_COUNT){
            
            let adIdAction = UIAlertAction(title: advTypeTextArray[i], style: .default, handler: {(UIAlertAction) in
                           self.placementIdTextField.placeholder = placeholderArray[i]
                           return;
                       })
                changePosIdController.addAction(adIdAction)
            }
        return changePosIdController
    }
    
    @IBAction func selectAdType(_ sender: UIButton) {
        let changePosIdController = configAlertController()
        self.present(changePosIdController, animated: false, completion: nil)
    
    }
    
//    func clicktoBackController(){
//
//        var backToMainView = UIView()
//        if let arrayViews = UIApplication.shared.keyWindow?.subviews{
//        for x in arrayViews {
//            let viewNameStr = NSString(format: "%s", object_getClassName(x))
//            if viewNameStr.isEqual(to: "UITransitionView"){
//            backToMainView = x.subviews[0]
//                break;
//            }
//        }
//        }
//        //    UIView *backToMainView = [arrayViews.lastObject subviews][0];
//        backToMainView.isUserInteractionEnabled = true
//        let mybackTap = UITapGestureRecognizer(target: self, action: #selector(backTap))
//        backToMainView.addGestureRecognizer(mybackTap)
//    }
//
//    @objc func backTap(){
//        self.changePosIdController.dismiss(animated: true, completion: nil)
//    }
    @IBAction func loadAd(_ sender: UIButton) {
        let placementId = self.placementIdTextField.text?.count ?? 0 > 0 ? self.placementIdTextField.text : self.placementIdTextField.placeholder
        self.nativeExpressAd = GDTNativeExpressAd(placementId: placementId, adSize: CGSize(width: widthSliderValue, height: heightSliderValue))
        self.nativeExpressAd.delegate = self
        self.nativeExpressAd.load(self.adCountSliderValue)
            }
    
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        self.view.endEditing(true)
    }
    
    //    Mark:UITableViewDelegate
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.row % 2 == 0 {
            let view: UIView = expressAdViews[indexPath.row/2]
            return view.bounds.size.height
        }
        return 44
    }
    
    //    MARK:UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return (expressAdViews != nil) ? expressAdViews.count * 2 : 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell:UITableViewCell!
        
        if indexPath.row % 2 == 0 {
            cell = tableView.dequeueReusableCell(withIdentifier: "nativeexpresscell", for: indexPath)
            cell.selectionStyle = .none
            
            let subView: UIView? = cell!.contentView.viewWithTag(1000)
            if (subView?.superview != nil) {
                subView?.removeFromSuperview()
            }
            
            let view: UIView = expressAdViews[indexPath.row / 2]
            view.tag = 1000
            cell.contentView.addSubview(view)
        } else {
            cell = tableView.dequeueReusableCell(withIdentifier: "splitnativeexpresscell", for: indexPath)
            cell.backgroundColor = .gray
        }
        
        return cell
    }
    
    //    Mark:GDTNativeExpressAdDelegete
    func nativeExpressAdSuccess(toLoad nativeExpressAd: GDTNativeExpressAd!, views: [GDTNativeExpressAdView]!) {
        expressAdViews = Array.init(views)
        if expressAdViews.count > 0 {
            for obj in expressAdViews {
                let expressView:GDTNativeExpressAdView = obj
                expressView.controller = self
                expressView.render()
            }
        }
        tableView.reloadData()
    }
    
    func nativeExpressAdFail(toLoad nativeExpressAd: GDTNativeExpressAd!, error: Error!) {
        print("Express Ad Load Fail : \(String(describing: error))")
    }
    
    func nativeExpressAdViewRenderSuccess(_ nativeExpressAdView: GDTNativeExpressAdView!) {
        tableView.reloadData()
    }
    
    func nativeExpressAdViewClicked(_ nativeExpressAdView: GDTNativeExpressAdView!) {
        print(#function)
    }
    
    func nativeExpressAdViewClosed(_ nativeExpressAdView: GDTNativeExpressAdView!) {
        print(#function)
        if let removeIndex = expressAdViews.index(of: nativeExpressAdView) {
            expressAdViews.remove(at: removeIndex)
            tableView.reloadData()
        }
    }
}
